### PR TITLE
autodetect default branch name for automated MRs

### PR DESCRIPTION
**Pull request by geoberle**
the a-i default branch name can't be assumed to be "master" for all a-i repos but can be "main" or anything else as well.

this PR introduces main branch name detection.

(also added some typehints along the way, thought it is not comprehensive since the gitlab api does not really expose good typehints on its own 😢 )

https://issues.redhat.com/browse/APPSRE-8420

### CODE REVIEW

File: reconcile/utils/gitlab_api.py [hunk 1]
```diff
@@ -100,6 +100,22 @@ def __init__(
...
+        return self._main_branch or "master"
+
+    def _determine_main_branch(self) -> Optional[str]:
+        if self.project:
+            main_branch_obj = next(
+                (b for b in self.project.branches.list() if b.default), None
+            )
+            if main_branch_obj:
+                return main_branch_obj.name
+        return None
```
Comment (hemslo) [inline reconcile/utils/gitlab_api.py:118]: It seems we want to cache `main_branch` computation, right?
How about using `cached_property`?

```python
    DEFAULT_MAIN_BRANCH = "master"

    @cached_property
    def main_branch(self) -> str:
        if self.project:
            return next(
                (b.name for b in self.project.branches.list() if b.default),
                DEFAULT_MAIN_BRANCH,
            )
        return DEFAULT_MAIN_BRANCH
```

**Review by hemslo — COMMENTED** [review]
To support more types, we need to update `PyGithub` version.

File: reconcile/utils/mr/base.py [hunk 2]
```diff
@@ -206,17 +206,17 @@ def delete_tmp_branch(self, gitlab_cli):
...
-    def submit(self, cli: MRClient):
+    def submit(self, cli: MRClient) -> Union[Any, None]:
         if isinstance(cli, GitLabApi):
             return self.submit_to_gitlab(gitlab_cli=cli)
 
         if isinstance(cli, SQSGateway):
-            return self.submit_to_sqs(sqs_cli=cli)
+            self.submit_to_sqs(sqs_cli=cli)
 
         raise AttributeError(f"client {cli} not supported")
```
Comment (chassing) [inline reconcile/utils/mr/base.py:221]: Why did you remove `return`?

File: reconcile/utils/gitlab_api.py (hunk 1 above)
Comment (geoberle) [reply reconcile/utils/gitlab_api.py:118]: there are some code paths where the projects becomes available only later. so i want to avoid caching a default branch name when there might be a better one on the project

File: reconcile/utils/mr/base.py (hunk 2 above)
Comment (geoberle) [reply reconcile/utils/mr/base.py:221]: `submit_to_sqs` does not return anything and mypy was starting to complain when i added the type hints

File: reconcile/utils/gitlab_api.py (hunk 1 above)
Comment (hemslo) [reply reconcile/utils/gitlab_api.py:118]: then how about this:

```python
    @cached_property
    def project_main_branch(self) -> str:
        return next(
            (b.name for b in self.project.branches.list() if b.default),
            DEFAULT_MAIN_BRANCH,
        )

    @property
    def main_branch(self) -> str:
        return this.project_main_branch if self.project else DEFAULT_MAIN_BRANCH
```

File: reconcile/utils/mr/base.py (hunk 2 above)
Comment (chassing) [reply reconcile/utils/mr/base.py:221]: But now, if someone runs it with `SQSGateway`, it will always raise an `AttributeError`.

File: reconcile/utils/gitlab_api.py (hunk 1 above)
Comment (geoberle) [reply reconcile/utils/gitlab_api.py:118]: i like this, i will use it.

in general i think the gitlab_api.py code needs some refactoring, because it mixes pure API things with projects things. being able to switch out the project behind the scene makes me kind of nervous.

in the meantime, your code will work nicely. thank you

**Review by chassing — APPROVED** [review]

**Review by hemslo — APPROVED** [review]
LGTM!
